captum/optim/_utils/reducer.py

```diff
@@ -0,0 +1,192 @@
+
+    @classmethod
+    def _apply_flat(cls, func: Callable, x: torch.Tensor) -> torch.Tensor:
+        orig_shape = x.shape
+        return func(x.reshape([-1, x.shape[-1]])).reshape(list(orig_shape[:-1]) + [-1])
+
+    def fit_transform(
+        self, x: Union[torch.Tensor, np.ndarray], reshape: bool = True
+    ) -> Union[torch.Tensor, np.ndarray]:
+        """
+        Perform dimensionality reduction on an input tensor or NumPy array.
+        """
+
+        is_tensor = torch.is_tensor(x)
+
+        if is_tensor:
+            if x.dim() == 3 and reshape:
+                x = x.permute(2, 1, 0)
+            elif x.dim() > 3 and reshape:
+                permute_vals = [0] + list(range(x.dim()))[2:] + [1]
+                x = x.permute(*permute_vals)
+        else:
+            if x.ndim == 3 and reshape:
+                x = x.transpose(2, 1, 0)
+            elif x.ndim > 3 and reshape:
+                permute_vals = [0] + list(range(x.ndim))[2:] + [1]
+                x = x.transpose(*permute_vals)
+
+        x_out = ChannelReducer._apply_flat(self._reducer.fit_transform, x)
+
+        if is_tensor:
+            x_out = torch.as_tensor(x_out, device=x.device)
+            if x.dim() == 3 and reshape:
+                x_out = x_out.permute(2, 1, 0)
+            elif x.dim() > 3 and reshape:
+                permute_vals = (
+                    [0]
+                    + [x.dim() - 1]
+                    + list(range(x.dim()))[1 : len(list(range(x.dim()))) - 1]
+                )
+                x_out = x_out.permute(*permute_vals)
+        else:
+            if x.ndim == 3 and reshape:
+                x_out = x_out.permute(2, 1, 0)
+            elif x.ndim > 3 and reshape:
+                permute_vals = (
+                    [0]
+                    + [x.ndim - 1]
+                    + list(range(x.ndim))[1 : len(list(range(x.ndim))) - 1]
+                )
+                x_out = x_out.transpose(*permute_vals)
+
+        return x_out
+
+    def __getattr__(self, name: str) -> Any:
+        if name in self.__dict__:
+            out = self.__dict__[name]
+        elif name + "_" in self._reducer.__dict__:
+            out = self._reducer.__dict__[name + "_"]
+        if type(out) == np.ndarray:
+            out = torch.as_tensor(out)
+        return out
+
+    def __dir__(self) -> List:
+        dynamic_attrs = [
+            name[:-1]
+            for name in dir(self._reducer)
+            if name[-1] == "_" and name[0] != "_"
+        ]
+
+        return (
+            list(ChannelReducer.__dict__.keys())
+            + list(self.__dict__.keys())
+            + dynamic_attrs
+        )
+
+
+def posneg(x: torch.Tensor, dim: int = 0) -> torch.Tensor:
+    """
+    Hack that makes a matrix positive by concatination in order to simulate
+    one-sided NMF with regular NMF
+    """
+
+    return torch.cat(
+        [torch.max(x, torch.full_like(x, 0)), torch.max(-x, torch.full_like(x, 0))],

```

Comment:
@NarineK Yeah, I guess it is the same! Lucid probably did it that way because TensorFlow doesn't have functional layers or anything like that. Should we keep is as it is, or use ReLU?

Reply:
@ProGamerGov, thank you for comparing with lucid. I think we can use ReLU since it is cleaner and probably easier to grasp.

Reply:
Okay, I've made the change!